File: tests/integration/fixtures/simple_service.js
/**
 * Simple payment processing service - Node.js version
 */

class User {
  constructor(id, name, email, balance) {
    this.id = id;
    this.name = name;
    this.email = email;
    this.balance = balance;
  }
}

class Transaction {
  constructor(transactionId, status, amount, userId) {
    this.transactionId = transactionId;
    this.status = status;
    this.amount = amount;
    this.userId = userId;
  }
}

class PaymentProcessor {
  constructor() {
    this.users = new Map();
    this.transactions = [];
  }

  getUser(userId) {
    return this.users.get(userId) || null;
  }

  addUser(user) {
    this.users.set(user.id, user);
  }

  processPayment(userId, amount) {
    const user = this.getUser(userId);

    if (user === null) {
      throw new Error(`User not found: ${userId}`);
    }

    if (user.balance < amount) {
      throw new Error(`Insufficient balance: ${user.balance} < ${amount}`);
    }

    // Update balance
    user.balance -= amount;

    // Create transaction
    const transaction = new Transaction(
      `tx_${this.transactions.length + 1}`,
      'completed',
      amount,
      userId
    );

    this.transactions.push(transaction);
    return transaction;
  }

  getBalance(userId) {
    const user = this.getUser(userId);
    if (user === null) {
      return 0.0;
    }
    return user.balance;
  }

  listTransactions(userId) {
    return this.transactions.filter(t => t.userId === userId);
  }
}

function calculateFee(amount) {
  if (amount < 10.0) {
    return 0.5;
  } else if (amount < 100.0) {
    return amount * 0.02;
  } else {
    return amount * 0.01;
  }
}

async function asyncValidatePayment(userId, amount) {
  // Simulate async validation
  if (amount <= 0) {
    return false;
  }
  return true;
}

module.exports = {
  User,
  Transaction,
  PaymentProcessor,
  calculateFee,
  asyncValidatePayment
};

File: tests/integration/fixtures/simple_service.go
// Simple payment processing service - Go version
package main

import (
	"errors"
	"fmt"
)

type User struct {
	ID      string
	Name    string
	Email   string
	Balance float64
}

type Transaction struct {
	TransactionID string
	Status        string
	Amount        float64
	UserID        string
}

type PaymentProcessor struct {
	users        map[string]*User
	transactions []Transaction
}

func NewPaymentProcessor() *PaymentProcessor {
	return &PaymentProcessor{
		users:        make(map[string]*User),
		transactions: make([]Transaction, 0),
	}
}

func (p *PaymentProcessor) GetUser(userID string) *User {
	return p.users[userID]
}

func (p *PaymentProcessor) AddUser(user *User) {
	p.users[user.ID] = user
}

func (p *PaymentProcessor) ProcessPayment(userID string, amount float64) (Transaction, error) {
	user := p.GetUser(userID)

	if user == nil {
		return Transaction{}, errors.New(fmt.Sprintf("User not found: %s", userID))
	}

	if user.Balance < amount {
		return Transaction{}, errors.New(fmt.Sprintf("Insufficient balance: %f < %f", user.Balance, amount))
	}

	// Update balance
	user.Balance -= amount

	// Create transaction
	transaction := Transaction{
		TransactionID: fmt.Sprintf("tx_%d", len(p.transactions)+1),
		Status:        "completed",
		Amount:        amount,
		UserID:        userID,
	}

	p.transactions = append(p.transactions, transaction)
	return transaction, nil
}

func (p *PaymentProcessor) GetBalance(userID string) float64 {
	user := p.GetUser(userID)
	if user == nil {
		return 0.0
	}
	return user.Balance
}

func (p *PaymentProcessor) ListTransactions(userID string) []Transaction {
	result := make([]Transaction, 0)
	for _, t := range p.transactions {
		if t.UserID == userID {
			result = append(result, t)
		}
	}
	return result
}

func CalculateFee(amount float64) float64 {
	if amount < 10.0 {
		return 0.5
	} else if amount < 100.0 {
		return amount * 0.02
	} else {
		return amount * 0.01
	}
}

func AsyncValidatePayment(userID string, amount float64) bool {
	// Simulate async validation
	if amount <= 0 {
		return false
	}
	return true
}
